### PR TITLE
Rewrite ItemCommand to C++

### PR DESCRIPTION
**Pull request by GnomedDev**

### CODE REVIEW

File: payload/sp/ItemCommand.cc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    // No args
+    if (!strcmp(args, "/i")) {
+        OSReport("&aClearing item\n");
+        ClearItem(myPlayerId);
+        return;
+    }
+
+    // Args
+    if (!strncmp(args, "/i ", strlen("/i "))) {
```
Comment (riidefi) [inline payload/sp/ItemCommand.cc:80]: Flip if statement

File: payload/sp/ItemCommand.cc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        u32 qty = 1;
+        if (!sscanf(args, "/i %d %d", &item, &qty)) {
+            OSReport("&aUnknown arguments \"%s\". Usage: /i [item] [qty|0=STICKY]\n",
+                    args + strlen("/i "));
+            return;
+        }
+        if (qty <= 0) {
+            auto &stickyState = s_sticky.emplace();
+
+            qty = MAX(-qty, 1);
```
Comment (riidefi) [inline payload/sp/ItemCommand.cc:91]: In C++ we can use std::max

File: payload/sp/ItemCommand.cc [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+}
+#include <game/system/RaceConfig.hh>
+
+extern "C" {
+#include <revolution.h>
+}
+
+namespace SP {
+
+s8 GetMyPlayerID(void) {
```
Comment (riidefi) [inline payload/sp/ItemCommand.cc:17]: `int` not `s8`

File: payload/sp/ItemCommand.cc [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    if (!strncmp(args, "/i ", strlen("/i "))) {
+        u32 item = 0;
+        u32 qty = 1;
+        if (!sscanf(args, "/i %d %d", &item, &qty)) {
+            OSReport("&aUnknown arguments \"%s\". Usage: /i [item] [qty|0=STICKY]\n",
+                    args + strlen("/i "));
+            return;
+        }
+        if (qty <= 0) {
+            auto &stickyState = s_sticky.emplace();
```
Comment (riidefi) [inline payload/sp/ItemCommand.cc:89]: s_sticky.emplace() = {
.item = ...,
.quantity = ...
};

File: payload/sp/ItemCommand.cc (hunk 3 above)
Comment (GnomedDev) [reply payload/sp/ItemCommand.cc:17]: Why? It isn't ever going to be above 127.

Comment (riidefi) [reply payload/sp/ItemCommand.cc:17]: You're right; I misread it as an implicit narrowing type conversion.

File: payload/sp/ItemCommand.cc [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        TrySet(myPlayerId, item, qty);
+        return;
+    }
+
+    OSReport("&4Invalid command: \"%s\"", args);
+}
+
+std::optional<ItemCommand::StickyState> ItemCommand::s_sticky = std::nullopt;
```
[the remark this answers is not in the document]
Comment (GnomedDev) [reply payload/sp/ItemCommand.cc:106]: I've seen it done both ways.

**Review by riidefi — APPROVED** [review]

File: payload/sp/ItemCommand.cc [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+}
+#include <game/system/RaceConfig.hh>
+
+extern "C" {
+#include <revolution.h>
+}
+
+namespace SP {
+
+s8 GetMyPlayerID(void) {
```
Comment (MikeIsAStar) [inline payload/sp/ItemCommand.cc:17]:
```suggestion
static s8 GetMyPlayerID() {
```

File: payload/sp/ItemCommand.cc [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    auto *raceConfig = System::RaceConfig::Instance();
+    if (raceConfig == nullptr) {
+        OSReport("&cNot in a race.\n");
+        return -1;
+    }
+
+    auto &raceScenario = raceConfig->raceScenario();
```
Comment (vabold) [inline payload/sp/ItemCommand.cc:24]:
```suggestion
    auto &raceScenario = System::RaceConfig::Instance()->raceScenario();
```
RaceConfig is preserved through all scenes and is created on game start.

File: payload/sp/ItemCommand.cc [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    s8 playerId = raceScenario.screenPlayerIds[0];
```
Comment (vabold) [inline payload/sp/ItemCommand.cc:25]: The error message suggests that the expected behavior is for the screen player ID to be a valid, real player. We should verify the type of player is real local or clarify the expected behavior in the error message.

**Review by vabold — CHANGES_REQUESTED** [review]

File: payload/sp/ItemCommand.cc (hunk 7 above)
Comment (GnomedDev) [reply payload/sp/ItemCommand.cc:24]: This code really was busted before I rewrote it.

Comment (GnomedDev) [reply payload/sp/ItemCommand.cc]: I'll reimplement an actual check after #803 allows me to get the SceneManager.

Comment (vabold) [reply payload/sp/ItemCommand.cc]: Actually, we can use the section ID to do this.

**Review by stblr — APPROVED** [review]

**Pull request merged**